k2/csrc/determinize.cc

```diff
@@ -7,44 +7,136 @@
+// arc that traces back to a previous LogSumTracebackState.
+// A LogSumTracebackState represents a weighted colletion of paths
+// terminating in a specific state.
+struct LogSumTracebackLink {
+
+  int32_t arc_index;  // Index of most recent arc in path from start-state to
+                      // the dest-state, or -1 if the path is empty (only
+                      // possible if this element belongs to the start-state).
+
+  int32_t symbol;     // Symbol on the arc numbered `arc_index` of the input FSA
+                      // (copied here for convenience), or 0 if arc_index == -1.
+
+  double prob;        // The probability mass associated with this incoming
+                      // arc in the LogSumTracebackState to which this belongs.
+
+  std::shared_ptr<LogSumTracebackState> prev_state;
+};
+
+struct LogSumTracebackState {
+  // LogSumTracebackState can be thought of as as a weighted set of paths from the
+  // start state to a particular state.  (It will be limited to the subset of
+  // paths that have a specific symbol sequence).
+
+  // `prev_elements` is, conceptually, a list of pairs (incoming arc-index,
+  // traceback link); we will keep it free of duplicates of the same incoming
+  // arc.
+  vector<LogSumTracebackLink> prev_elements;
+
+
+  int32_t arc_index;  // Index of most recent arc in path from start-state to
+                      // the dest-state, or -1 if the path is empty (only
+                      // possible if this element belongs to the start-state).
+
+  int32_t symbol;     // Symbol on the arc numbered `arc_index` of the input FSA
+                      // (copied here for convenience), or 0 if arc_index == -1.
+
+  MaxTracebackState(std::shared_ptr<MaxTracebackState> prev,
+              int32_t arc_index, int32_t symbol):
+      prev(prev), arc_index(arc_index), symbol(symbol) { }
+
+};
+
+
+struct DetStateElement {
+
+  double weight;      // Weight from reference state to this state, along
+                      // the path taken by following the 'prev' links
+                      // (the path would have `seq_len` arcs in it).
+                      // Note: by "this state" we mean the destination-state of
+                      // the arc at `arc_index`.
+                      // Interpret this with caution, because the
+                      // base state, and the length of the sequence arcs from the
+                      // base state to here, are known only in the DetState
+                      // that owns this DetStateElement.
+
+  std::shared_ptr<PathElement> path;
+                      // The path from the start state to here (actually we will
+                      // only follow back `seq_len` links.  Will be nullptr if
+                      // seq_len == 0 and this belongs to the initial determinized
+                      // state.
+
+  DetStateElement &&Advance(float arc_weight, int32_t arc_index, int32_t arc_symbol) {
+    return DetStateElement(weight + arc_weight,
```

Comment:
just a question, we always do `+` here, right? Suppose it it log-sum?

Reply:
It's always + because this is along a path.  It's when we merge paths that it's max vs. log-sum.

Reply:
you can fix this later, BTW, let's not block merge on this.